src/linear_algebra/vandermonde.lean

```diff
@@ -0,0 +1,313 @@
+      by_cases hi : i = 0,
+      { rw [if_pos hi, if_pos hi, if_pos hi, mul_comm, ←neg_mul_eq_neg_mul,
+          (pow_succ _ _).symm, hkmp, mul_one, add_left_neg] },
+      { rw [if_neg hi, if_neg hi, if_neg hi, mul_one, mul_comm, ←neg_mul_eq_neg_mul],
+        nth_rewrite 1 ←hkmp,
+        rw geom_sum₂_succ_eq,
+        ring }}},
+end
+
+lemma fin_one_sum_fin_congr_equiv_symm_apply_right {j : fin (n + 1)} (h : j ≠ 0) :
+  (fin_sum_fin_equiv.trans (fin_congr (add_comm 1 n))).symm j = sum.inr (j.pred h) :=
+begin
+  simp only [fin_congr_symm, symm_trans_apply],
+  erw fin_congr_apply_mk (add_comm n 1) _ j.2,
+  rw fin_sum_fin_equiv_symm_apply_right,
+  { simp only [],
+    apply fin.eq_of_veq,
+    simp only [fin.val_eq_coe, fin.coe_pred] },
+  { exact nat.pos_of_ne_zero (λ hj, h (fin.eq_of_veq hj)) }
+end
+
+lemma vandermonde_step3_det' :
+  (vandermonde_step3 R n a).det = (vandermonde R n (λ i, a (i + 1))).det :=
+begin
+  transitivity
+    (matrix.from_blocks (1 : matrix (fin 1) (fin 1) R) 0 0 (vandermonde R n (λ i, a (i + 1)))).det,
+  { convert det_reindex_self' (equiv.trans fin_sum_fin_equiv (fin_congr (add_comm 1 n)))
+      (matrix.from_blocks (1 : matrix (fin 1) (fin 1) R) 0 0 (vandermonde R n (λ i, a (i + 1)))),
+    dunfold vandermonde_step3,
+    ext i j,
+    by_cases hi : i = 0,
+    { rw hi,
+      by_cases hj : j = 0,
+      { simp [hj] },
+      { rw [if_neg hj, fin_one_sum_fin_congr_equiv_symm_apply_right _ hj],
+        simp }},
+    { rw [if_neg hi, fin_one_sum_fin_congr_equiv_symm_apply_right _ hi],
+      by_cases hj : j = 0,
+      { rw hj,
+        simp },
+      { rw [if_neg hj, fin_one_sum_fin_congr_equiv_symm_apply_right _ hj,
+          from_blocks_apply₂₂, vandermonde_def, fin.coe_pred _ hi, fin.coe_pred _ hj,
+          nat.sub_add_cancel],
+        exact nat.pos_of_ne_zero (λ hi', hi (fin.eq_of_veq hi')) }}},
+  rw upper_two_block_triangular_det,
+  simp
+end
+
+lemma vandermonde_det :
+  det (vandermonde R n a) = ∏ i in range n, ∏ j in range i, ((a i) - (a j)) :=
+begin
+  induction n with n hn generalizing a,
+  { apply matrix.det_eq_one_of_card_eq_zero (fintype.card_fin 0) },
+  { rw [vandermonde_step2_det, vandermonde_step3_det, vandermonde_step3_det',
+      hn (λ (i : ℕ), a (i + 1))],
+    simp only [],
+    nth_rewrite 0 prod_range_succ',
+    rw [range_zero, finset.prod_empty, mul_one],
+    have hr : ∀ (k : ℕ), ∏ (j : ℕ) in range (k + 1), (a (k + 1) - a j) =
+      (∏ (j : ℕ) in range k, (a (k + 1) - a (j + 1))) * (a (k + 1) - a 0),
+    { intro k, apply prod_range_succ' },
+    simp_rw hr,
+    rw [prod_mul_distrib, mul_comm] }
+end
+
+/-- Definition of a Vandermonde matrix, using a : fin (n + 1) → R -/
+def vandermonde_fin (a : fin n → R) : matrix (fin n) (fin n) R :=
+λ i j, (a i) ^ (j : ℕ)
+
+lemma vandermonde_fin_det (a : fin n → R) : det (vandermonde_fin R n a) =
+  ∏ i : fin n, ∏ j : {j : fin n // j < i}, ((a i) - (a j)) :=
```

Comment:
I'm surprised this proof is so long, since it should be trivial on paper to derive it from `vandermonde_det`. Wouldn't something like the following work:
 * rewrite `vandermonde_fin R n a = vandermonde R n (a \comp coe)`
 * apply `vandermonde_det`
 * rewrite `∏ j : {j : fin n // j < i}, g ↑j = ∏ j in range ↑i, g j` and `∏ i : fin n, f ↑i = ∏ i in range n, f i`.